File: sdk/go/hovel/chain_kv.go
package hovel

import (
	"errors"
	"fmt"
	"strings"
)

var ErrChainKVUnavailable = errors.New("hovel: chain kv is not available in this runtime")

type ChainKVBinding struct {
	Key         string `json:"key"`
	ConfigKey   string `json:"configKey,omitempty"`
	StepID      string `json:"stepId,omitempty"`
	Description string `json:"description,omitempty"`
	Required    bool   `json:"required,omitempty"`
}

type ChainKVContract struct {
	Requires []ChainKVBinding `json:"requires,omitempty"`
	Produces []ChainKVBinding `json:"produces,omitempty"`
}

// ChainKVContractProvider is optional. Existing modules do not need to
// implement it; modules use it to declare validated survey-to-exploit handoffs.
type ChainKVContractProvider interface {
	ChainKVContract() ChainKVContract
}

type ChainKVResolution struct {
	Value  any
	Source string
	Found  bool
}

type chainKVMutation struct {
	Operation string `json:"operation"`
	Key       string `json:"key"`
	Value     string `json:"value,omitempty"`
}

type ChainKV struct {
	available bool
	target    string
	revision  uint64
	entries   map[string]string
	mutations []chainKVMutation
}

func newChainKV(target string, revision uint64, entries map[string]string) *ChainKV {
	cloned := make(map[string]string, len(entries))
	for key, value := range entries {
		cloned[key] = value
	}
	return &ChainKV{available: true, target: target, revision: revision, entries: cloned}
}

func unavailableChainKV(target string) *ChainKV {
	return &ChainKV{target: target, entries: map[string]string{}}
}

func (k *ChainKV) Available() bool { return k != nil && k.available }

func (k *ChainKV) Revision() uint64 {
	if k == nil {
		return 0
	}
	return k.revision
}

func (k *ChainKV) Get(key string) (string, bool) {
	if k == nil || !k.available {
		return "", false
	}
	value, ok := k.entries[k.expand(key)]
	return value, ok
}

func (k *ChainKV) Exists(key string) bool {
	_, ok := k.Get(key)
	return ok
}

func (k *ChainKV) Set(key, value string) error {
	if !k.Available() {
		return ErrChainKVUnavailable
	}
	key = k.expand(key)
	if strings.TrimSpace(key) == "" {
		return errors.New("hovel: chain kv key is required")
	}
	k.entries[key] = value
	k.mutations = append(k.mutations, chainKVMutation{Operation: "set", Key: key, Value: value})
	return nil
}

func (k *ChainKV) Delete(key string) error {
	if !k.Available() {
		return ErrChainKVUnavailable
	}
	key = k.expand(key)
	if strings.TrimSpace(key) == "" {
		return errors.New("hovel: chain kv key is required")
	}
	delete(k.entries, key)
	k.mutations = append(k.mutations, chainKVMutation{Operation: "delete", Key: key})
	return nil
}

func (k *ChainKV) expand(key string) string {
	return strings.ReplaceAll(key, "{target}", percentEncodeChainKVTarget(k.target))
}

func percentEncodeChainKVTarget(value string) string {
	var out strings.Builder
	for _, b := range []byte(value) {
		if b >= 'a' && b <= 'z' || b >= 'A' && b <= 'Z' || b >= '0' && b <= '9' || strings.ContainsRune("-_.~", rune(b)) {
			out.WriteByte(b)
		} else {
			_, _ = fmt.Fprintf(&out, "%%%02X", b)
		}
	}
	return out.String()
}

func (k *ChainKV) wireMutations() map[string]any {
	if k == nil || !k.available || len(k.mutations) == 0 {
		return nil
	}
	return map[string]any{"baseRevision": k.revision, "operations": append([]chainKVMutation(nil), k.mutations...)}
}

func (c *Context) ChainKV() *ChainKV {
	if c.chainKV == nil {
		c.chainKV = unavailableChainKV(c.Target)
	}
	return c.chainKV
}

func (c *Context) ResolveInput(configKey, kvKey string, def any) ChainKVResolution {
	if value, ok := c.Inputs[configKey]; ok {
		return ChainKVResolution{Value: value, Source: "input", Found: true}
	}
	if value, ok := c.TargetConfig[configKey]; ok {
		return ChainKVResolution{Value: value, Source: "target-config", Found: true}
	}
	if value, ok := c.ChainConfig[configKey]; ok {
		return ChainKVResolution{Value: value, Source: "chain-config", Found: true}
	}
	if value, ok := c.ChainKV().Get(kvKey); ok {
		return ChainKVResolution{Value: value, Source: "chain-kv", Found: true}
	}
	return ChainKVResolution{Value: def, Source: "default", Found: false}
}

File: sdk/rust/hovel/src/chain_kv.rs
use crate::json::Value;

#[derive(Clone, Debug, Default)]
pub struct ChainKVBinding {
    pub key: String,
    pub config_key: String,
    pub step_id: String,
    pub description: String,
    pub required: bool,
}

impl ChainKVBinding {
    pub fn new(key: &str) -> ChainKVBinding {
        ChainKVBinding {
            key: key.to_string(),
            ..Default::default()
        }
    }

    fn to_value(&self) -> Value {
        Value::object(vec![
            ("key", Value::from(self.key.as_str())),
            ("configKey", Value::from(self.config_key.as_str())),
            ("stepId", Value::from(self.step_id.as_str())),
            ("description", Value::from(self.description.as_str())),
            ("required", Value::Bool(self.required)),
        ])
    }
}

#[derive(Clone, Debug, Default)]
pub struct ChainKVContract {
    pub requires: Vec<ChainKVBinding>,
    pub produces: Vec<ChainKVBinding>,
}

impl ChainKVContract {
    pub(crate) fn to_value(&self) -> Value {
        Value::object(vec![
            (
                "requires",
                Value::Array(self.requires.iter().map(ChainKVBinding::to_value).collect()),
            ),
            (
                "produces",
                Value::Array(self.produces.iter().map(ChainKVBinding::to_value).collect()),
            ),
        ])
    }
}

#[derive(Clone, Debug)]
pub struct ChainKV {
    available: bool,
    target: String,
    revision: u64,
    entries: Vec<(String, String)>,
    operations: Vec<Value>,
}

impl ChainKV {
    pub(crate) fn from_params(target: &str, params: &Value) -> ChainKV {
        let payload = params.get("chainKV");
        let available = matches!(payload, Some(Value::Object(_)));
        let revision = payload
            .and_then(|value| value.get("revision"))
            .and_then(Value::as_f64)
            .unwrap_or(0.0) as u64;
        let entries = payload
            .and_then(|value| value.get("entries"))
            .and_then(Value::as_object)
            .unwrap_or(&[])
            .iter()
            .filter_map(|(key, value)| value.as_str().map(|text| (key.clone(), text.to_string())))
            .collect();
        ChainKV {
            available,
            target: target.to_string(),
            revision,
            entries,
            operations: Vec::new(),
        }
    }

    pub fn available(&self) -> bool {
        self.available
    }
    pub fn revision(&self) -> u64 {
        self.revision
    }

    pub fn get(&self, key: &str) -> Option<&str> {
        let key = self.expand(key);
        self.entries
            .iter()
            .find(|(candidate, _)| candidate == &key)
            .map(|(_, value)| value.as_str())
    }

    pub fn exists(&self, key: &str) -> bool {
        self.get(key).is_some()
    }

    pub fn set(&mut self, key: &str, value: &str) -> Result<(), String> {
        if !self.available {
            return Err("hovel: chain kv is not available in this runtime".into());
        }
        let key = self.expand(key);
        if key.trim().is_empty() {
            return Err("hovel: chain kv key is required".into());
        }
        if let Some((_, current)) = self
            .entries
            .iter_mut()
            .find(|(candidate, _)| candidate == &key)
        {
            *current = value.to_string();
        } else {
            self.entries.push((key.clone(), value.to_string()));
        }
        self.operations.push(Value::object(vec![
            ("operation", Value::from("set")),
            ("key", Value::Str(key)),
            ("value", Value::from(value)),
        ]));
        Ok(())
    }

    pub fn delete(&mut self, key: &str) -> Result<(), String> {
        if !self.available {
            return Err("hovel: chain kv is not available in this runtime".into());
        }
        let key = self.expand(key);
        if key.trim().is_empty() {
            return Err("hovel: chain kv key is required".into());
        }
        self.entries.retain(|(candidate, _)| candidate != &key);
        self.operations.push(Value::object(vec![
            ("operation", Value::from("delete")),
            ("key", Value::Str(key)),
        ]));
        Ok(())
    }

    pub(crate) fn mutations(&self) -> Option<Value> {
        if !self.available || self.operations.is_empty() {
            return None;
        }
        Some(Value::object(vec![
            ("baseRevision", Value::Num(self.revision as f64)),
            ("operations", Value::Array(self.operations.clone())),
        ]))
    }

    fn expand(&self, key: &str) -> String {
        key.replace("{target}", &percent_encode(&self.target))
    }
}

fn percent_encode(value: &str) -> String {
    let mut out = String::new();
    for byte in value.as_bytes() {
        if byte.is_ascii_alphanumeric() || matches!(*byte, b'-' | b'_' | b'.' | b'~') {
            out.push(*byte as char);
        } else {
            out.push_str(&format!("%{byte:02X}"));
        }
    }
    out
}
